src/QsFmt/SPECIFICATION.md

```diff
@@ -0,0 +1,58 @@
+﻿# Q# Formatter Design Specification Document
+
+QsFmt is a source code formatter for Q#.
+
+## Where to get the formatter
+
+The `Microsoft.Quantum.QsFormatter` NuGet package will be distributed with the rest of Q# on NuGet.org.
+As a dotnet tool, it is installed with the command `dotnet tool install Microsoft.Quantum.QsFormatter`.
+
+## How to use the command line formatter
+
+After installing the tool the command `qsfmt` is used to run the formatter.
+There are two commands supported for the tool: 'format' and 'update'.
+The 'format' command runs the tool as a formatter for Q# code, running only those transformation rules that update whitespace and affect indentation. The underlying code will not be changed.
+The 'update' command allows the tool to be used as a means of updating old Q# code, replacing deprecated syntax with newer supported syntax. These transformation rules change the actual code given to them, and may occasionally fail.
+These two commands work separately and can't be run together. If the user wishes to both format and update their code, they would need to run the tool twice, first with 'update' and then with 'format'.
+
+The tool will initially only support the 'update' command, as the formatting functionality is still being worked on. The 'format' command will be unavailable until it is supported.
+
+## Input and Output
+Input files to the formatter will be the last argument(s) to the tool. Zero to many arguments may be given.
+ - The default behavior, which is used if no input arguments are provided, is for the formatter to take as input all files with the '.qs' extension in the current working directory.
+ - Inputs can be given as one to many paths to directories, in which all files with the '.qs' extension found will be taken as input.
+ - Inputs can be given as one to many paths to individual files with the '.qs' extension.
+ - Inputs can be any combination of paths to files and paths to directories.
+ - The `--recurse` option can be specified to process the input directories recursively for input files. You can specify this for all input directories, not per-directory.
+
+The output of the formatter is to overwrite the input files that it processes.
+
```

Comment:
Will the formatter provide the transaction semantics (either successful completion or no change to the file)? 
Or, if the formatter updates the input file on the fly, and hits an error or crashes, then what happens to the input file? Will the input file remain partially overwritten (or removed/corrupted because its file handle has not been closed properly)?

Reply:
Ideally if an error or unhandled exception is encountered, the file currently being process does not get overwritten. The file would only get overwritten if the formatting was successful. I suspect the behavior should be different for updating errors, though I still don't really understand what that means. @samarsha can you better explain what an error means in the context of an update transformation?

Reply:
For example, `new Qubit[5]` has no equivalent in the new array syntax, because there is no default value for `Qubit`. It can't be updated automatically without refactoring the code.

Reply:
@samarsha so what should be the expected response from the tool? It shouldn't transform those instances, but does it need to do anything else?

Reply:
I don't think that has been decided yet, but I think displaying a warning for every instance that it couldn't update would make sense.

Reply:
Alright, thanks @samarsha. Let's put a pin in this for now, and come back to it once it becomes more relevant.